overlord/snapstate/check_snap_test.go

```diff
@@ -680,3 +685,237 @@ base: some-base
 	st.Lock()
 	c.Check(err, IsNil)
 }
+
+func (s *checkSnapSuite) TestValidateContainerReallyEmptyFails(c *C) {
+	const yaml = `name: empty-snap
+version: 1
+`
+	d := c.MkDir()
+	info, err := snap.InfoFromSnapYaml([]byte(yaml))
+	c.Assert(err, IsNil)
+
+	err = snapstate.ValidateContainer(info, snapdir.New(d))
+	c.Check(err, Equals, snapstate.ErrMissingPaths)
+}
+
+func (s *checkSnapSuite) TestValidateContainerEmptyButBadPermFails(c *C) {
+	const yaml = `name: empty-snap
+version: 1
+`
+	d := c.MkDir()
+
+	stat, err := os.Stat(d)
+	c.Assert(err, IsNil)
+	c.Check(stat.Mode().Perm(), Equals, os.FileMode(0700)) // just to be sure
+
+	c.Assert(os.Mkdir(filepath.Join(d, "meta"), 0755), IsNil)
+	c.Assert(ioutil.WriteFile(filepath.Join(d, "meta", "snap.yaml"), nil, 0444), IsNil)
+
+	info, err := snap.InfoFromSnapYaml([]byte(yaml))
+	c.Assert(err, IsNil)
+
+	err = snapstate.ValidateContainer(info, snapdir.New(d))
+	c.Check(err, Equals, snapstate.ErrBadModes)
+}
+
+func (s *checkSnapSuite) TestValidateContainerMissingSnapYamlFails(c *C) {
+	const yaml = `name: empty-snap
+version: 1
+`
+	d := c.MkDir()
+	c.Assert(os.Chmod(d, 0755), IsNil)
+	c.Assert(os.Mkdir(filepath.Join(d, "meta"), 0755), IsNil)
+
+	info, err := snap.InfoFromSnapYaml([]byte(yaml))
+	c.Assert(err, IsNil)
+
+	err = snapstate.ValidateContainer(info, snapdir.New(d))
+	c.Check(err, Equals, snapstate.ErrMissingPaths)
+}
+
+func (s *checkSnapSuite) TestValidateContainerSnapYamlBadPermsFails(c *C) {
```

Comment:
Pardon my ignorance again, but how is this different from TestValidateContainerEmptyButBadPermFails ? Looks quite similar to me, maybe a comment that highlights the differences? For if it is about permissions of "." vs "meta" vs "meta/snap.yaml" - maybe make it table driven?

Reply:
i'll take a look

Reply:
while the error and yaml could be tabled, the main differences are in the little setups. Maybe I should add a comment explaining what it's doing? highlighting differences